File: Data Structures & Algorithms/word-break/submission-1.java
class Solution {
    Boolean[] dp;

    public boolean helper(int i, String s, Set<String> wordSet){
        int n = s.length();
        if(i == n)
            return true;
        if(dp[i] != null)
            return dp[i];
        for(int j = i + 1; j <= n; j++){
            if(wordSet.contains(s.substring(i, j)) && helper(j, s, wordSet)){
                dp[i] = true;
                return dp[i];
            }
        }
        dp[i] = false;
        return dp[i];
    }

    public boolean wordBreak(String s, List<String> wordDict) {
        int n = s.length();
        dp = new Boolean[n + 1];
        Set<String> wordSet = new HashSet<>(wordDict);
        return helper(0, s, wordSet);
    }
}

File: Data Structures & Algorithms/count-paths/submission-1.py
class Solution:
    def uniquePaths(self, m: int, n: int) -> int:
        dp = [[0] * (n + 1) for _ in range(m + 1)]
        for i in range(1, m + 1):
            dp[i][1] = 1
        for j in range(1, n + 1):
            dp[1][j] = 1
        for i in range(2, m + 1):
            for j in range(2, n + 1):
                dp[i][j] = dp[i - 1][j] + dp[i][j - 1]
        return dp[-1][-1]
        

'''
if m == 0 and n == 0:
    return 0
if m == 1 and n == 1:
    return 1

helper(i, j) = helper(i - 1, j) + helper(i, j - 1)
'''